Create conditionals.md

ZThatWhiteGuyZ opened the pull request:
Basic skeleton of the tutorial. Pushing because the weather is bad.

Commit:
Create conditionals.md

Basic skeleton of the tutorial. Pushing because the weather is bad.

Commit:
Update conditionals.md

Worked through the if and else section.  Hit on anything non-zero is considered true, conditionals are built using logical operators to make boolean expressions, and basic structure of if and else statements.

Commit:
Update conditionals.md

Else if tutorial written. Ternary and Short Circuiting left, and possibly a deeper dive into multi-condition conditionals.

Commit:
Update conditionals.md

Worked on ternary operator tutorial. Looking forward to that review. Added a short circuiting heading.

Commit:
Update conditionals.md

Wrote the tutorial on multiple-condition statements and short circuiting. Included plenty of examples, and even some code that is meant to be copied and run by students to tangibly see short circuiting. Multi-condition Statements feels a little short. Might also have some weird wording.

Commit:
Update conditionals.md

Removed comment at the top, and added comment on the bottom. No change to tutorial text.

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
````diff
@@ -0,0 +1,284 @@
+}
+```
+
+In this example, sunny is being compared to true. If it is sunny, then the condition will be met, and we will execute `close_the_blinds()`. If it is not sunny (`sunny = 0`), then the condition statement would essentially be `0 == 1`, which we know is false. This would mean that `close_the_blinds()` would not be executed.
````
I would add tick marks around "sunny" since it is used as a variable. But only the first "sunny"

```suggestion
In this example, `sunny` is being compared to true. If it is sunny, then the condition will be met, and we will execute `close_the_blinds()`. If it is not sunny (`sunny = 0`), then the condition statement would essentially be `0 == 1`, which we know is false. This would mean that `close_the_blinds()` would not be executed.
```

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
```diff
@@ -0,0 +1,284 @@
+
+---
+
+`else` is the jelly to the `if` peanut butter. Not required, but it makes it better. So, to jelly up our example, if it is sunny then I will close the blinds. Otherwise, it its \[raining, cloudy, chance of meatballs, snowy, dark\] and the blinds will be open. `else` is a catch-all for things that make the condition false. The basic structure would look something like this:
```
"its" -> "is"

```suggestion
`else` is the jelly to the `if` peanut butter. Not required, but it makes it better. So, to jelly up our example, if it is sunny then I will close the blinds. Otherwise, it is \[raining, cloudy, chance of meatballs, snowy, dark\] and the blinds will be open. `else` is a catch-all for things that make the condition false. The basic structure would look something like this:
```

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
```diff
@@ -0,0 +1,284 @@
+}
+else
+{
+	expression_2
```
Missing semicolon

```suggestion
	expression_2;
```

ZThatWhiteGuyZ replied:
Self review: mirror ternary example expression_is_true and _false in if-else.

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
````diff
@@ -0,0 +1,284 @@
+The ternary operator is named so because it takes three operands.  It is similar to an `if else` in that it takes in a condition statement and then has an outcome for true and false. Looking at the structure will help.
+
+``` c
+condition ? expression_1 : expression_2;
````
I may just rewrite this to be clear which one is for true and which one is for false, but since the next few lines show it it isn't super required.

```suggestion
condition ? expression_if_true : expression_if_false;
```

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
````diff
@@ -0,0 +1,284 @@
+}
+```
+
+Here is a different example with a condition that compares two integers:
+
+``` c
+#include <stdio.h>
+
+int main(void)
+{
+	int num1 = 20, num2 = 23;
+
+	(num1 > num2) ? printf("%d is greater than %d\n", num1, num2) : printf("%d is greater than %d\n", num2, num1);
+
+	return 0;
+}
+```
````
I like this example, but could you add the output of this program? Just for illustration.

ZThatWhiteGuyZ replied:
Yes.

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
````diff
@@ -0,0 +1,284 @@
+}
+```
+
+Let's try it with the ternary operator.
+
+``` c
+#include <stdio.h>
+
+int main(void)
+{
+	int m = 20, n = 15, g = 17, p = 3;
+
+	(m > n && m > g && m > p) ? printf("%d is the greatest.\n", m) : printf("%d is not the greatest.\n", m);
+
+return 0;
+}
+```
````
Could you also add the output here too? This condition is extra funky.

ZThatWhiteGuyZ replied:
Yes.

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
```diff
@@ -0,0 +1,284 @@
+
+---
+
+When writing conditional statements, it is good practice to write them in a way that allows for the compiler to skip the evaluation of further conditions when possible. This skip by the compiler is called short circuiting. Short circuiting is possible when evaluating logical `and` and logical `or`. For example, if the left side of a logical and is false, the whole expression will be false regardl ...
```
Putting tick marks around this one "and"
```suggestion
When writing conditional statements, it is good practice to write them in a way that allows for the compiler to skip the evaluation of further conditions when possible. This skip by the compiler is called short circuiting. Short circuiting is possible when evaluating logical `and` and logical `or`. For example, if the left side of a logical `and` is false, the whole expression will be false regardless of the right side of the expression, so the compiler skips this evaluation. Here is an example of such a case:
```

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
```diff
@@ -0,0 +1,284 @@
+
+`x` and `y` are not equivalent, so the comparison would resolve to false, rendering the whole statement false.
+
+With a slight change, here is a similar example for a logical or short circuit:
```
Ticks around "or"
```suggestion
With a slight change, here is a similar example for a logical `or` short circuit:
```

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
````diff
@@ -0,0 +1,284 @@
+}
+```
+
+Note that logical or will resolved if **either** side of the expression is true, so comparing `y` and `z` will resolve to true, rendering the whole statement true.
````
```suggestion
Note that logical `or` will resolved if **either** side of the expression is true, so comparing `y` and `z` will resolve to true, rendering the whole statement true.
```

jerrettl replied:
I would add an extra sentence or two here just to flesh out this wording. Felt like it was just missing something.
```suggestion
Note that logical or will resolved if **either** side of the expression is true, so comparing `y` and `z` will resolve to true, rendering the whole statement true. Since `y == z` is true, `x` and `y` don't even need to be compared. The result of `x == y` would have no effect on the truthiness of the logical expression.
```

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
````diff
@@ -0,0 +1,284 @@
+}
+```
+
+When you run this program, the output will show that `x = 1`, even though it was incremented in the in the if statement. This part of the code was actually never executed due to the short circuit, so `x` was never incremented. This code can be slightly manipulated to show the short circuiting of the logical `and` as well, but I will leave that for you to try.
````
`x` was not actually incremented so I was confused by the first sentence until I read the next sentence. This just adds a little hint for the reader.

```suggestion
When you run this program, the output will show that `x = 1`, even though it was (supposedly) incremented in the in the if statement. This part of the code was actually never executed due to the short circuit, so `x` was never incremented. This code can be slightly manipulated to show the short circuiting of the logical `and` as well, but I will leave that for you to try.
```

jerrettl replied:
in the in the

```suggestion
When you run this program, the output will show that `x = 1`, even though it was incremented in the if statement. This part of the code was actually never executed due to the short circuit, so `x` was never incremented. This code can be slightly manipulated to show the short circuiting of the logical `and` as well, but I will leave that for you to try.
```

jerrettl reviewed (commented):
Looks really nice. I like this a lot. I saw just a few little typos here and there, and made some typesetting suggestions. Ternary operators were written well, but I suggested an extra sentence or two to clarify a part. It felt to me like that one part was a little short. Otherwise looks good. Charge on

ZThatWhiteGuyZ committed:
Apply suggestions from code review

nice suggs

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

Commit:
Update conditionals.md

Added periods to the code on line 160, as well as added in the old suggestions that couldn't be committed.

Commit:
Update conditionals.md

Added that extra piece in the last paragraph of short circuiting about switching the condition order for different behavior. Could use a check on that. Next section is a work in progress, but will accept feedback.

Commit:
Update conditionals.md

Finished up on one-line if and else statements. Happy to add anything else to the conditionals tutorial if needed.

Commit:
Update conditionals.md

Changed the title to match the practice.

jerrettl commented on src/intro-to-c/conditionals/conditionals.md:
````diff
@@ -0,0 +1,332 @@
+	printf("Nothing is the same!\n");
+```
+
+This comes in handy when dealing with many short conditionals. However, there is an important pitfall to be aware of. C will only recognize the next following executable line as within the `if` statement. That is, even if you put twenty new lines between `if` and the "Something is the same" `printf`, it will be executed if the condition is true. If you did something like this, though, your compil ...
````
```suggestion
This comes in handy when dealing with many short conditionals. However, there is an important pitfall to be aware of. C will only recognize the next following executable line as within the `if` statement. That is, even if you put twenty new lines between `if` and `printf("Something is the same!\n");`, "Something is the same!" will still only be executed if the condition is true. If you did something like this, though, your compiler will be unhappy:
```

ZThatWhiteGuyZ replied:
I am down with this sickness.

jerrettl reviewed (commented):
Checked out your new changes. Everything looks good, but I just suggested one sentence that I thought could be rephrased. Let me know what you think.

ZThatWhiteGuyZ committed:
Update src/intro-to-c/conditionals/conditionals.md

Co-authored-by: Jerrett Longworth <jerrett7@gmail.com>

jerrettl reviewed (approved):
🤝

ZThatWhiteGuyZ commented:
If there is nothing else you think this needs, then I'm good with this being merged.

jerrettl commented:
Do it 👀
